staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -578,11 +596,17 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
-	klog.V(1).Info("[graceful-termination] RunPreShutdownHooks has completed")
 
 	// Wait for all requests in flight to drain, bounded by the RequestTimeout variable.
 	<-drainedCh.Signaled()
+
+	if s.AuditBackend != nil {
+		s.AuditBackend.Shutdown()
+		klog.V(1).InfoS("[graceful-termination] audit backend shutdown completed")
+	}
+
 	// wait for stoppedCh that is closed when the graceful termination (server.Shutdown) is finished.
+	<-listenerStoppedCh
 	<-stoppedCh
```

Comment:
these channels are returned on L533 by ` s.NonBlockingRun(`
```
	stoppedCh, listenerStoppedCh, err := s.NonBlockingRun(stopHttpServerCh, shutdownTimeout)
	if err != nil {
		return err
	}
```
is there any reasons to wait for both now?

Reply:
depends on how busy the server is, but I think it's the opposite - `listenerStoppedCh` should fire first since the first thing `Server.Shutdown` does is close the listener and it causes the `Server.Serve` to return
https://github.com/golang/go/blob/master/src/net/http/server.go#L2771-L2775

and this consequently causes `listenerStoppedCh` to be closed
https://github.com/kubernetes/kubernetes/blob/78a4ba6af8c24b7d0e0efabfdbfed72fe4fa00ec/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go#L240-L250

and `stoppedCh` is closed once `Server.Shutdown` is done waiting for the existing socket connection to go idle
https://github.com/golang/go/blob/7ce38641564e5c0183bb09df8c523caf5e0ab8e1/src/net/http/server.go#L2796-L2806

https://github.com/kubernetes/kubernetes/blob/78a4ba6af8c24b7d0e0efabfdbfed72fe4fa00ec/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go#L232-L238

Reply:
but yeah, i think the order in which we wait is not important, but i am fine if you still want me to reverse it.